src/main/scala/firrtl/backends/experimental/smt/TransitionSystem.scala

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: Apache-2.0
+// Author: Kevin Laeufer <laeufer@cs.berkeley.edu>
+
+package firrtl.backends.experimental.smt
+
+import firrtl.graph.MutableDiGraph
+import scala.collection.mutable
+
+private case class State(sym: SMTSymbol, init: Option[SMTExpr], next: Option[SMTExpr]) {
+  def name: String = sym.name
+}
+private case class Signal(name: String, e: SMTExpr, lbl: SignalLabel = IsNode) {
+  def toSymbol: SMTSymbol = SMTSymbol.fromExpr(name, e)
+  def sym:      SMTSymbol = toSymbol
+}
+private case class TransitionSystem(
+  name:     String,
+  inputs:   List[BVSymbol],
+  states:   List[State],
+  signals:  List[Signal],
+  comments: Map[String, String] = Map(),
+  header:   String = "") {
+  def serialize: String = TransitionSystem.serialize(this)
+}
+
+private sealed trait SignalLabel
+private case object IsNode extends SignalLabel
+private case object IsOutput extends SignalLabel
+private case object IsConstraint extends SignalLabel
+private case object IsBad extends SignalLabel
+private case object IsFair extends SignalLabel
+private case object IsNext extends SignalLabel
+private case object IsInit extends SignalLabel
+
+private object SignalLabel {
+  private val labels = Seq(IsNode, IsOutput, IsConstraint, IsBad, IsFair, IsNext, IsInit)
+  val labelStrings = Seq("node", "output", "constraint", "bad", "fair", "next", "init")
+  val labelToString: SignalLabel => String = labels.zip(labelStrings).toMap
+  val stringToLabel: String => SignalLabel = labelStrings.zip(labels).toMap
+}
+
+private object TransitionSystem {
+  def serialize(sys: TransitionSystem): String = {
+    (Iterator(sys.name) ++
+      sys.inputs.map(i => s"input ${i.name} : ${SMTExpr.serializeType(i)}") ++
+      sys.signals.map(s => s"${SignalLabel.labelToString(s.lbl)} ${s.name} : ${SMTExpr.serializeType(s.e)} = ${s.e}") ++
+      sys.states.map(serialize)).mkString("\n")
+  }
+
+  def serialize(s: State): String = {
+    s"state ${s.sym.name} : ${SMTExpr.serializeType(s.sym)}" +
+      (s.init match {
+        case None       => ""
+        case Some(init) => s"\n  [init] ${init}"
+      }) +
+      (s.next match {
+        case None       => ""
+        case Some(next) => s"\n  [next] ${next}"
+      })
```

Comment:
```suggestion
      s.init.map("\n  [init] " + _).getOrElse("") +
      s.next.map("\n  [next] " + _).getOrElse("")
```
Bit of a nit, but IMO it's more clear to avoid concatenating the results of multi-line expressions.

Reply:
this is an awesome catch!